eZ/Publish/Core/Persistence/Legacy/Content/Location/Handler.php

```diff
@@ -310,20 +310,55 @@ public function copySubtree($sourceId, $destinationParentId)
             );
         }
 
-        // If subtree root is main location for its content, update subtree section to the one of the
-        // parent location content
-        $subtreeRootContentInfo = $this->contentHandler->loadContentInfo($copiedSubtreeRootLocation->contentId);
-        if ($subtreeRootContentInfo->mainLocationId === $copiedSubtreeRootLocation->id) {
-            $this->setSectionForSubtree(
-                $copiedSubtreeRootLocation->id,
-                $this->contentHandler->loadContentInfo($this->load($destinationParentId)->contentId)->sectionId
-            );
-        }
+        $destinationParentSectionId = $this->getSectionId($destinationParentId);
+        $this->updateSubtreeSectionIfNecessary($copiedSubtreeRootLocation, $destinationParentSectionId);
 
         return $copiedSubtreeRootLocation;
     }
 
     /**
+     * Retrieves section ID of the location's content.
+     *
+     * @param int $locationId
+     *
+     * @return int
+     */
+    private function getSectionId($locationId)
+    {
+        $location = $this->load($locationId);
+        $locationContentInfo = $this->contentHandler->loadContentInfo($location->contentId);
+
+        return $locationContentInfo->sectionId;
+    }
+
+    /**
+     * If the location is the main location for its content, updates subtree section.
+     *
+     * @param Location $location
+     * @param int $sectionId
+     */
+    private function updateSubtreeSectionIfNecessary(Location $location, $sectionId)
+    {
+        if ($this->isMainLocation($location)) {
+            $this->setSectionForSubtree($location->id, $sectionId);
```

Comment:
What I meant is that we should ensure the section is set for the whole tree, as you say - but I was uncertain if it would make sense performance wise to only set it on the content which didn't have the same section set already. This would require checking the whole subtree, which may be slow - versus not checking and always recursively set section, which may also be slow (and is a waste of time if section does not actually change). Which is worse? That would be a statistical question of how often the section actually changes, not easy to answer. I wanted to mention it, in case others had anything clever to add.

On the other hand, not checking section is the alternative with the smallest/simplest code, that's a real plus. I'm okay with it for that reason.

> In the legacy, it actually did, but only for the source node, not all of its children.

Right, so legacy produces somewhat arbitrary results. That might be good to document.